addons/website_slides/static/src/js/public/components/slide_upload_dialog/slide_upload_category.js

```diff
@@ -0,0 +1,441 @@
+            url: this.state.form.url,
+            slide_category: this.props.slideCategory,
+            channel_id: this.props.channelId,
+        });
+        this.canSubmitForm = true;
+        if (data.error) {
+            this._alertDisplay(data.error);
+            this.state.preview.show = false;
+        } else {
+            if (data.info) {
+                this._alertDisplay(data.info, "alert-info");
+            } else {
+                this._alertRemove();
+            }
+
+            this.isValidUrl = true;
+
+            if (data.name) {
+                this.state.form.slideName = data.name;
+                this.state.preview.videoTitle = data.name;
+                this.state.preview.hideSlideVideoTitle = false;
+            } else {
+                this.state.preview.hideSlideVideoTitle = true;
+            }
+
+            if (data.completion_time) {
+                // hours to minutes conversion
+                this.state.form.duration = Math.round(data.completion_time * 60);
+            }
+            if (data.image_url) {
+                this.state.form.slideImage = data.image_url;
+            }
+
+            if (!data.name && !data.image_url) {
+                this.state.preview.show = false;
+            } else {
+                this.state.preview.show = true;
+            }
+        }
+
+        this.state.form.isLoading = false;
+    }
+
+    async onClickFormSubmit(forcePublished) {
+        if (!this._formValidate()) {
+            return;
+        }
+        const values = await this._formValidateGetValues(forcePublished);
+        this.props.upload(values, this.props.slideCategory);
+    }
+
+    /**
+     * When the user selects 'local_file' or 'external' as source type, we display the 'upload'
+     * field or the 'document_google_url' / 'image_google_url' fields respectively.
+     */
+    onClickSourceType(isLocalSource) {
+        this.state.form.isLocalSource = isLocalSource;
+    }
+
+    //--------------------------------------------------------------------------
+    // Private
+    //--------------------------------------------------------------------------
+
+    // Alert messages
+
+    /**
+     * @param {string} message
+     */
+    _alertDisplay(message, alertClass = "alert-warning") {
+        this.state.alert.message = message;
+        this.state.alert.class = alertClass;
+        this.state.alert.show = true;
+    }
+
+    _alertRemove() {
+        this.state.alert.show = false;
+        this.state.alert.message = "";
+        this.state.alert.class = "";
+    }
+
+    // Category and tag SelectMenus
+
+    /**
+     * Get value for category_id and tag_ids (ORM cmd) to send to server
+     */
+    _getSelectMenuValues() {
+        const result = {};
+        // tags
+        if (this.state.choices.tagIds.length > 0) {
+            result["tag_ids"] = this.state.choices.tagIds.map((tagId) => {
+                const tag = this.state.choices.tags.find((tag) => tag.value === tagId);
```

Comment:
O(n^2) complexity, can we build a dict before we call map ?
and so in the map we can do:
```js
tag = tags[tagId]